src/java.base/share/classes/jdk/internal/util/HexDigits.java

```diff
@@ -91,6 +91,19 @@ public final class HexDigits implements Digits {
     private HexDigits() {
     }
 
+    /**
+     * For values from 0 to 256 return a short encoding a pair of hex ASCII-encoded digit characters in little-endian
+     * @param i value to convert
+     * @param ucase ture uppper case, false lower case
+     * @return a short encoding a pair of hex ASCII-encoded digit characters
+     */
+    public static short digitPair(int i, boolean ucase) {
+        short v = DIGITS[i & 0xff];
+        return ucase
+                ? (short) (v - ((v & 0b0100_0000_0100_0000) >> 1)) // really: v - ((v >= 'a' && v <= 'f') ? 32 : 0)
```

Comment:
I think this logic is somewhat complicated that explaining directly is better:
`0b0100_0000_0100_0000` is a selector that selects letters (1 << 6), uppercase or not, and shifting it right by 1 bit incidentally becomes a bit offset between cases (1 << 5).

I think you can keep the original `& ~` which is clearer (as `-` will not work had the differences not be aligned bit-wise). But you should explain what `>> 1` does which is a major hurdle to understanding this function.

Reply:
If we changed DIGITS to be encoded with the uppercase digits then the expression could be simplified to
```return ucase ? v : (short) (v | 0b0010_0000_0010_0000); // or 0x2020```

Reply:
Some performance-focused scenarios, such as UUID.toString, use lowercase, so I think DIGITS should be lowercase.

Reply:
> I think this logic is somewhat complicated that explaining directly is better: `0b0100_0000_0100_0000` is a selector that selects letters (1 << 6), uppercase or not, and shifting it right by 1 bit incidentally becomes a bit offset between cases (1 << 5).
> 
> I think you can keep the original `& ~` which is clearer (as `-` will not work had the differences not be aligned bit-wise). But you should explain what `>> 1` does which is a major hurdle to understanding this function.

I have added your explanation and it is very well written.

I changed it to "-" because there is one less operation and it is easier to explain.